app/code/Magento/Inventory/Test/Integration/Stock/IsProductInStockTest.php

```diff
@@ -0,0 +1,139 @@
+        /** @var IndexStructureInterface $indexStructure */
+        $indexStructure = Bootstrap::getObjectManager()->get(IndexStructureInterface::class);
+
+        foreach ([1, 2, 3] as $stockId) {
+            $indexName = $indexNameBuilder
+                ->setIndexId(StockItemIndexerInterface::INDEXER_ID)
+                ->addDimension('stock_', $stockId)
+                ->setAlias(Alias::ALIAS_MAIN)
+                ->create();
+            $indexStructure->delete($indexName);
+        }
+
+        // Cleanup reservations
+        $this->reservationCleanup->execute();
+    }
+
+    /**
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/products.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stocks.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stock_source_link.php
+     */
+    public function testProductIsInStock()
+    {
+        $this->indexer->reindexRow(1);
+
+        $reservations = [
+            $this->reservationBuilder->setStockId(1)->setSku('SKU-1')->setQuantity(-5)->build(), // reserve 5 units
+            $this->reservationBuilder->setStockId(1)->setSku('SKU-1')->setQuantity(1.5)->build(), // unreserve 1.5 units
+        ];
+        $this->reservationsAppend->execute($reservations);
+
+        self::assertTrue($this->isProductInStockService->execute('SKU-1', 1));
+
+        $reservations = [
+            $this->reservationBuilder->setStockId(1)->setSku('SKU-1')->setQuantity(3.5)->build(), // unreserve 3.5 units
+        ];
+        $this->reservationsAppend->execute($reservations);
+    }
+
+    /**
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/products.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stocks.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stock_source_link.php
+     */
+    public function testProductIsNotInStock()
+    {
+        $this->indexer->reindexRow(1);
+
+        $reservations = [
+            $this->reservationBuilder->setStockId(1)->setSku('SKU-1')->setQuantity(-8.5)->build(), // reserve 8.5 units
+        ];
+        $this->reservationsAppend->execute($reservations);
+
+        self::assertFalse($this->isProductInStockService->execute('SKU-1', 1));
+
+        $reservations = [
+            $this->reservationBuilder->setStockId(1)->setSku('SKU-1')->setQuantity(8.5)->build(), // unreserve 8.5 units
+        ];
+        $this->reservationsAppend->execute($reservations);
```

Comment:
is this kind of rollback action? which should prevent to affect further calculations?